File: leet/1129_shortest_path_with_alternating_colors/1129_shortest_path_with_alternating_colors.ts
// LeetCode 1129. Shortest Path with Alternating Colors
// https://leetcode.com/problems/shortest-path-with-alternating-colors/
// O(E+V) time, O(E+V) space
function shortestAlternatingPaths(n: number, red_edges: number[][], blue_edges: number[][]): number[] {
    enum Edge {
        Red = "reds",
        Blue = "blues"
    }
    const edges = {reds: {}, blues: {}};
    for (let [from, to] of red_edges) {
        if (!edges.reds[from])
            edges.reds[from] = new Set();
        edges.reds[from].add(to);
    }
    for (let [from, to] of blue_edges) {
        if (!edges.blues[from])
            edges.blues[from] = new Set();
        edges.blues[from].add(to);
    }
    const visited = {
        reds: new Array(n).fill(-1), blues: new Array(n).fill(-1) };
    const dfs = function(i, color, steps) {
        if (visited[color][i] != -1 && steps >= visited[color][i])
            return;
        visited[color][i] = steps;
        if (!edges[color][i])
            return;
        edges[color][i].forEach((edge)=>(
            dfs(edge, color === Edge.Red ? Edge.Blue : Edge.Red, steps + 1)));
    }
    
    dfs(0, Edge.Red, 0);
    dfs(0, Edge.Blue, 0);
    return visited.blues.map((b, i) =>
        b == -1 || visited.reds[i] == -1
            ? b == -1 ? visited.reds[i] : b
            : Math.min(b, visited.reds[i])
        );
};

File: leet/1129_shortest_path_with_alternating_colors/1129_shortest_path_with_alternating_colors.py
# LeetCode 1129. Shortest Path with Alternating Colors
# https://leetcode.com/problems/shortest-path-with-alternating-colors/
# O(E+V) time, O(E+V) space
class Solution:
    def shortestAlternatingPaths(self, n: int, red_edges: List[List[int]], blue_edges: List[List[int]]) -> List[int]:
        redEdges, blueEdges = defaultdict(set), defaultdict(set)
        [redEdges[a].add(b) for (a, b) in red_edges]
        [blueEdges[a].add(b) for (a, b) in blue_edges]
        redVisited = [float(inf)] * n
        blueVisited = [float(inf)] * n
        
        def dfs(i, isRed, steps):
            visited = redVisited if isRed else blueVisited
            edges = redEdges if isRed else blueEdges
            if steps >= visited[i]:
                return
            visited[i] = steps
            for edge in edges[i]:
                dfs(edge, not isRed, steps + 1)
        
        dfs(0, True, 0)
        dfs(0, False, 0)
        return [min(r, b) if min(r, b) < float('inf') else -1 for r, b in zip(redVisited, blueVisited)]
